crates/nostr/src/nips/nip26.rs

```diff
@@ -43,73 +107,441 @@ pub fn sign_delegation(
+    pub fn get_signature(&self) -> Signature {
+        self.signature
+    }
+
+    /// Create a delegation tag (including the signature).
+    /// See also validate().
+    pub fn create(
+        delegator_keys: &Keys,
+        delegatee_pubkey: XOnlyPublicKey,
+        conditions_string: &str,
+    ) -> Result<Self, Error> {
+        let signature = sign_delegation(
+            delegator_keys,
+            delegatee_pubkey,
+            conditions_string.to_string(),
+        )?;
+        let conditions = Conditions::from_str(conditions_string)?;
+        Ok(Self {
+            delegator_pubkey: delegator_keys.public_key(),
+            conditions,
+            signature,
+        })
+    }
+
+    /// Validate a delegation tag, check signature and conditions.
+    pub fn validate(
+        &self,
+        delegatee_pubkey: XOnlyPublicKey,
+        event_properties: &EventProperties,
+    ) -> Result<(), Error> {
+        // verify signature
+        if let Err(_e) = verify_delegation_signature(
+            &self.get_delegator_pubkey(),
+            &self.get_signature(),
+            delegatee_pubkey,
+            self.get_conditions_string(),
+        ) {
+            return Err(Error::ConditionsValidation(
+                ValidationError::InvalidSignature,
+            ));
+        }
+
+        // validate conditions
+        self.conditions.evaluate(event_properties)?;
+
+        Ok(())
+    }
+
+    /// Convert to JSON string.
+    pub fn to_json(&self) -> Result<String, Error> {
+        let delegator_pubkey_hex = self.delegator_pubkey.to_string();
+        let tag = json!([
+            DELEGATION_KEYWORD,
+            delegator_pubkey_hex,
+            self.conditions.to_string(),
+            self.signature.to_string(),
+        ]);
+        let s = tag.to_string();
+        Ok(s)
+    }
+
+    /// Parse from a JSON string
+    pub fn from_json(s: &str) -> Result<Self, Error> {
+        let v = serde_json::from_str::<Value>(s)?;
+        let arr = match v.as_array() {
+            None => return Err(Error::DelegationTagParse),
+            Some(a) => a,
+        };
+        if arr.len() != 4 {
+            return Err(Error::DelegationTagParse);
+        }
+        if arr[0].as_str().unwrap_or("") != DELEGATION_KEYWORD {
+            return Err(Error::DelegationTagParse);
+        }
+        let delegator_pubkey = XOnlyPublicKey::from_str(arr[1].as_str().unwrap_or(""))?;
+        let conditions = Conditions::from_str(arr[2].as_str().unwrap_or(""))?;
+        let signature = Signature::from_str(arr[3].as_str().unwrap_or(""))?;
+        Ok(DelegationTag {
+            delegator_pubkey,
+            conditions,
+            signature,
+        })
+    }
+}
+
+impl fmt::Display for DelegationTag {
+    /// Return tag in JSON string format
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.to_json() {
+            Err(e) => write!(f, "(error {e})"),
+            Ok(s) => write!(f, "{s}"),
+        }
+    }
```

Comment:
```rust
write!(f, "{}", self.as_json())
```

Reply:
I changed as_json() to return String, so this could be also simplified.